compiler/noirc_evaluator/src/ssa/opt/pure.rs

```diff
@@ -100,14 +118,24 @@ impl std::fmt::Display for Purity {
 
 impl Function {
     fn is_pure(&self) -> (Purity, BTreeSet<FunctionId>) {
+        // Note, this function must be allowed to complete despite the fact that once the function is marked as impure
+        // then its final purity is known. This is because we need to collect all of the dependencies of the function
+        // to ensure that they are processed.
+        //
+        // This can be relaxed if we calculate the callgraph separately.
```

Comment:
cc @vezenovm as you're centralising how we generate callgraphs.

Reply:
This is referring more to how we collect the function dependencies and when `is_pure` should be called. We ultimately use our `CallGraph` in this pass still.